[native] Add support for flow control to PrestoExchangeSource

mbasmanova opened the pull request:
PrestoExchangeSource now participates in Exchange flow control by implementing
new request API:

``` 
velox::ContinueFuture request(uint32_t maxBytes) override; 
```

PrestoExchangeSource limits the size of the response by passing 'maxBytes' to
upstream worker (it used to hard-code this value to 32MB). This helps control 
total amount of data queued in ExchangeQueue.

PrestoExchangeSource::request() returns a future that completes when successful
possibly empty response has been received. PrestoExchangeSource no longer
retries on empty responses, but rather waits until ExchangeClient
calls 'request()' again. This allows ExchangeClient to limit the number of sources 
that are fetching data concurrently.

See https://github.com/facebookincubator/velox/issues/6005

```
== NO RELEASE NOTE ==
```

xiaoxmeng commented on presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp:
```diff
@@ -111,30 +111,40 @@ bool PrestoExchangeSource::shouldRequestLocked() {
 
-void PrestoExchangeSource::request() {
+velox::ContinueFuture PrestoExchangeSource::request(uint32_t maxBytes) {
+  auto [promise, future] =
```
Do we add a CHECK here promise_ is empty?

xiaoxmeng commented on presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp:
```diff
@@ -232,31 +243,31 @@ void PrestoExchangeSource::processDataResponse(
-      if (complete || !empty) {
-        requestPending_ = false;
-      }
+      requestPromise = std::move(promise_);
```
Add check here that promise_ is not empty? Thanks!

nit: requestPromise(promise_) doesn't need to be handled under the queue lock? Thanks!

xiaoxmeng commented on presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp:
```diff
@@ -293,6 +302,7 @@ void PrestoExchangeSource::processDataError(
           path,
           error),
       queue_);
+  promise_.setValue();
```
Could we add CHECK promise_ is not empty?

xiaoxmeng commented on presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp:
```diff
@@ -350,6 +360,9 @@ void PrestoExchangeSource::abortResults() {
 
 void PrestoExchangeSource::close() {
   closed_.store(true);
+  if (promise_.valid() && !promise_.isFulfilled()) {
```
exchange close is called by operator close and is there any potential race condition between async triggered request and the driver close? Do we need source lock to protecdt the update or access to request promise? Thanks!

xiaoxmeng reviewed (commented):
@mbasmanova thanks for the change!

mbasmanova commented:
@xiaoxmeng Meng, thank you for review. Indeed, there was a race between close() and async request processing. I updated the code to protect access to promise_ with queue's mutex, added checks and comments. Would you take another look?

xiaoxmeng commented on presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp:
```diff
@@ -232,56 +256,57 @@ void PrestoExchangeSource::processDataResponse(
+      requestPromise.setValue();
+    } else {
+      // The source must have been closed.
+      VELOX_CHECK(closed_.load());
```
Nice. Thanks!
```
nit: not sure if default memory order is good enough VELOX_CHECK(closed_);
```

xiaoxmeng reviewed (approved):
@mbasmanova LGTM. Thanks for the update!
